packages/next-international/src/app/client/create-i18n-provider-client.tsx

```diff
@@ -12,13 +12,20 @@ type I18nProviderWrapperProps = {
   children: ReactNode;
 };
 
+export const localesCache = new Map<string, Record<string, unknown>>();
+
 export function createI18nProviderClient<Locale extends BaseLocale>(
   I18nClientContext: Context<LocaleContext<Locale> | null>,
   locales: ImportedLocales,
   fallbackLocale?: Record<string, unknown>,
 ) {
   function I18nProvider({ locale, children }: I18nProviderProps) {
-    const { default: clientLocale } = use(locales[locale as keyof typeof locales]());
+    let clientLocale: any = localesCache.get(locale);
+
+    if (!clientLocale) {
+      clientLocale = use(locales[locale as keyof typeof locales]()).default;
+      localesCache.set(locale, clientLocale);
+    }
 
     const value = useMemo(
```

Comment:
I think the `flattenLocales` calls should be avoided here. `fallbackLanguage` could be computed earlier, and the clientLocale may be flatten on the `set` function call via a `Proxy`.

I think this useMemo should be as primitive/pure as possible, like:
```ts
const value = useMemo(
  () => ({
    localeContent: clientLocale,
    fallbackLocale,
    locale: locale as string
  }),
  [locale, clientLocale]
);
```

Reply:
I think we'll have more work to do with `flattenLocale()` to dedupe and call it less often. It's used in many places in both the App & Pages lib. We definitely want to tackle this in another PR.